include/exec/at_coroutine_exit.hpp

```diff
@@ -0,0 +1,256 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+// The original idea is taken from libunifex and adapted to stdexec.
+
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <exception>
+#include <type_traits>
+
+#include "stdexec/execution.hpp"
+#include "exec/task.hpp"
+
+namespace exec {
+namespace __at_coroutine_exit {
+struct __die_on_stop_t {
+  template <class _Receiver>
+    struct __receiver_id {
+      struct __t {
+        using __id = __receiver_id;
+        _Receiver __receiver_;
+
+        template <class... _Args>
+          friend void tag_invoke(stdexec::set_value_t, __t&& __self,
+                                _Args&&... __args) noexcept {
+            if constexpr (stdexec::__nothrow_callable<stdexec::set_value_t, _Receiver&&, _Args&&...>) {
+              stdexec::set_value((_Receiver &&) __self.__receiver_, (_Args &&) __args...);
+            } else {
+              try {
+                stdexec::set_value((_Receiver &&) __self.__receiver_, (_Args &&) __args...);
+              } catch (...) {
+                stdexec::set_error((_Receiver &&) __self.__receiver_, std::current_exception());
+              }
+            }
+          }
+
+        template <class _Error>
+          friend void tag_invoke(stdexec::set_error_t, __t&& __self,
+                                _Error&& __error) noexcept {
+            stdexec::set_error((_Receiver &&) __self.__receiver_, (_Error &&) __error);
+          }
+
+        [[noreturn]] friend void tag_invoke(stdexec::set_stopped_t, __t&&) noexcept {
+          std::terminate();
+        }
+
+        friend auto tag_invoke(stdexec::get_env_t, const __t& __self) noexcept {
+          return stdexec::get_env(__self.__receiver_);
+        }
+      };
+    };
+  template <class _R> using __receiver = stdexec::__t<__receiver_id<_R>>;
+
+  template <class _Sig> struct __return_type {
+    using type = _Sig;
+  };
+
+  template <class _R, class... _Args> struct __return_type<_R(_Args...)> {
+    using type = _R;
+  };
+
+  template <class _Sig>
+  using __return_type_t = typename __return_type<_Sig>::type;
+
+  template <class _Tag, class _With = stdexec::__> struct __replace_tag {
+    template <class _Arg>
+    using __f = stdexec::__if_c<std::is_same_v<__return_type_t<_Arg>, _Tag>, _With, _Arg>;
```

Comment:
```suggestion
    using __f = stdexec::__if<std::is_same<__return_type_t<_Arg>, _Tag>, _With, _Arg>;
```

The metaprogramming facilities will avoid instantiating the `std::is_same` class template if you write it this way.